node_list.go

```diff
@@ -1,46 +1,82 @@
 package memcacheha
 
+import "sync"
+
 // NodeList represents a list of memcache servers configured/discovered by this client.
 type NodeList struct {
-	Nodes map[string]*Node
+	lock  sync.RWMutex // on nodes
+	nodes map[string]*Node
 }
 
 // NewNodeList returns a new, empty NodeList
 func NewNodeList() *NodeList {
 	return &NodeList{
-		Nodes: map[string]*Node{},
+		lock:  sync.RWMutex{},
+		nodes: map[string]*Node{},
 	}
 }
 
 // GetHealthyNodes returns a map of config endpoints to Nodes where the node IsHealthy is true
 func (nodeList *NodeList) GetHealthyNodes() map[string]*Node {
 	out := map[string]*Node{}
-	for _, node := range nodeList.Nodes {
-		if node.IsHealthy {
+	nodeList.lock.RLock()
+	for _, node := range nodeList.nodes {
+		if node.IsHealthy() {
 			out[node.Endpoint] = node
 		}
 	}
+	nodeList.lock.RUnlock()
 	return out
 }
 
 // GetHealthyNodeCount returns the count of Nodes where the node IsHealthy is true
 func (nodeList *NodeList) GetHealthyNodeCount() int {
 	healthy := 0
-	for _, node := range nodeList.Nodes {
-		if node.IsHealthy {
+	nodeList.lock.RLock()
+	for _, node := range nodeList.nodes {
+		if node.IsHealthy() {
 			healthy++
 		}
 	}
+	nodeList.lock.RUnlock()
 	return healthy
 }
 
 // Exists returns true if a node for the given endpoint exists
 func (nodeList *NodeList) Exists(nodeAddr string) bool {
-	_, found := nodeList.Nodes[nodeAddr]
+	nodeList.lock.RLock()
+	_, found := nodeList.nodes[nodeAddr]
+	nodeList.lock.RUnlock()
 	return found
 }
 
 // Add the given node to this list
 func (nodeList *NodeList) Add(node *Node) {
-	nodeList.Nodes[node.Endpoint] = node
+	nodeList.lock.Lock()
+	nodeList.nodes[node.Endpoint] = node
+	nodeList.lock.Unlock()
+}
+
+func (nodeList *NodeList) SetNodes(nodeAddrs map[string]bool) []string {
+	removedAddrs := []string{}
```

Comment:
Shouldn't this add the new nodes - or replace with the old map with the new?

Reply:
This mirrors the functionality moved from `client.go`. Basically, `nodeAddrs` contains all node addresses that are running now. The nodes have already been added at `client.go:459`. This removes the nodes that are not running from the list. I'm not fully sure why this is needed (if a node is disconnected, why isn't it removed then and there from `NodeList`?), but this is functionally the same as before.

The function could have a better name to reflect what it is doing.

Reply:
`RemoveNodesExcept`?  Then you get to say `removedAddrs := client.Nodes.RemoveNodesExcept(incomingNodes)`.  It is an odd construction.  But I agree your change matches what it replaces.